[quant][graphmode] Add quantize_per_tensor.tensors

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#35916 [quant][graphmode] Add quantize_per_tensor.tensors**
* #35894 [quant][graphmode] Insert Observers for dynamic LSTM
* #35893 [quant][graphmode] Add new tensorlist observer for LSTM

Summary:
quantize_per_tensor can now accept list of tensors.
Needed for operators like LSTM and cat

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20830388](https://our.internmc.facebook.com/intern/diff/D20830388)

supriyar committed:
[quant][graphmode] Add quantize_per_tensor.tensors
Summary:
quantize_per_tensor can now accept list of tensors.
Needed for operators like LSTM and cat

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented on aten/src/ATen/native/quantized/QTensor.cpp:
```diff
@@ -18,6 +18,22 @@ Tensor quantize_per_tensor_cpu(
   return quantizer->quantize(self);
 }
 
+std::vector<Tensor> quantize_per_tensor_list_cpu(
```
@jerryzh168 just confirming that this is the right place to define it? I seem to be hitting linker errors.

jerryzh168 replied:
I think so

jerryzh168 commented on aten/src/ATen/native/quantized/QTensor.cpp:
```diff
@@ -18,6 +18,22 @@ Tensor quantize_per_tensor_cpu(
+    ScalarType dtype) {
+  std::vector<Tensor> quantized_tensors;
+  for (auto i = 0; i < tensors.size(); ++i) {
+    quantized_tensors.push_back(quantize_per_tensor_cpu(
```
should we call aten native function here?

supriyar replied:
Either way works I think. That function calls `quantize_per_tensor_cpu`

supriyar committed:
Update on "[quant][graphmode] Add quantize_per_tensor.tensors"


Summary:
quantize_per_tensor can now accept list of tensors.
Needed for operators like LSTM and cat

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
you might need to do similar changes to the python lint files as I did in https://github.com/pytorch/pytorch/pull/34348

jerryzh168 reviewed (approved):
Please wait for CI, I think this will break the type_hint tests

supriyar committed:
Update on "[quant][graphmode] Add quantize_per_tensor.tensors"


Summary:
quantize_per_tensor can now accept list of tensors.
Needed for operators like LSTM and cat

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]
